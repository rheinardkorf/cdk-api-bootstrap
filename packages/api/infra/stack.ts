import * as cdk from '@aws-cdk/core';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import * as lambda from '@aws-cdk/aws-lambda';
import * as yaml from 'yamljs';
import * as path from 'path';
import * as fs from 'fs';

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || "dev";
const stackId = `${wsConfig.stackPrefix}API-${stage}`;
const tableName = `${wsConfig.defaultTable}-${stage}`;

export class ServiceStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Get reference to DynamoDB table.
    // Must specify `globalIndexes` for permissions to work on indexes.
    const table = dynamodb.Table.fromTableAttributes(this,`${stackId}_DynamoDB`, {
        tableName,
        globalIndexes: ['GSI1', 'GSI2', 'GSI3']
    });

    // Create API Gateway.
    const api = new apigateway.RestApi(this, `${stackId}_API`, {
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowHeaders:['Content-Type', 'X-Amz-Date', 'X-Amz-Security-Token', 'Authorization', 'X-Api-Key', 'X-Requested-With', 'Accept', 'Access-Control-Allow-Methods', 'Access-Control-Allow-Origin', 'Access-Control-Allow-Headers']
      }
    });

    // Create common layer for include dependencies in all lambdas.
    const commonLayer = new lambda.LayerVersion(this, `${stackId}_Common`, {
        code: new lambda.AssetCode(`layers/common`),
        compatibleRuntimes: [lambda.Runtime.NODEJS_14_X]
    })

    // Custom Authorizer Function.
    const lambdaAuthorizerFunction = new lambda.Function(this, `${stackId}_Authorization_Handler`, {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: new lambda.AssetCode(`authorizer`),
      handler: `authorizer.authorizer`,
      layers: [commonLayer]
    })

    // Custom Authorizer.
    const lambdaAuthorizer = new apigateway.TokenAuthorizer(this, `${stackId}_TokenAuthorizer`, {
      handler: lambdaAuthorizerFunction
    })

    // Load the API Spec.
    const spec = yaml.load(path.resolve(__dirname, '../api-spec.yaml'))

    // Generate resources for API paths and methods.
    for ( const key in spec.paths ) {

      // New API Gateway resource for path.
      const resource = api.root.resourceForPath(key);

      // Get the root of the path to point to functions.
      const fullPath = key.replace(/^\//,'')
      const root = fullPath.split('/')[0]

      // For each of the path methods add a new API Gateway method.
      for ( const method in spec.paths[key]) {
        const pathMethod = spec.paths[key][method]
        const isSecured = pathMethod.security && true || false

        // Create a new Lambda Function to handle the method.
        const lambdaHandler = new lambda.Function(this, `${pathMethod.operationId}_Handler`, {
          runtime: lambda.Runtime.NODEJS_14_X,
          code: new lambda.AssetCode(`functions/${root}`),
          handler: `${pathMethod.operationId}.${pathMethod.operationId}`,
					layers: [commonLayer],
          environment: {
            TABLE_NAME: table.tableName
          },
          
        })

        if ( 'GET' === method.toUpperCase() ) {
          table.grantReadData(lambdaHandler);
        } else {
          table.grantReadWriteData(lambdaHandler);
        }

        const authorizer = isSecured && {
          authorizationType: apigateway.AuthorizationType.CUSTOM,
          authorizer: lambdaAuthorizer
        } || {};

        // Attach the Lambda function to the API Gateway resource.
        resource.addMethod(method.toUpperCase(), new apigateway.LambdaIntegration(lambdaHandler), {
          ...authorizer
        });
      }
    }
  }
}