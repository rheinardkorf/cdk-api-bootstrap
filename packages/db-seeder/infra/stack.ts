import * as cdk from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cr from '@aws-cdk/custom-resources';
import * as logs from '@aws-cdk/aws-logs';
import * as iam from '@aws-cdk/aws-iam';
import * as path from 'path'
import * as fs from 'fs'

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || "dev";
const stackId = `${wsConfig.stackPrefix}DB-Seeder-${stage}`;
const tableName = `${wsConfig.defaultTable}-${stage}`;
const seedsStage = process.env.SEEDS_STAGE || stage;

export class SeederStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table: dynamodb.ITable = dynamodb.Table.fromTableAttributes(this,`${stackId}__DynamoDB`, {
        tableName,
        globalIndexes: ['GSI1', 'GSI2', 'GSI3']
    });

    const seederResourceRole = new iam.Role(this,`${stackId}_seederResourceRole`,{
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com')
    })

    seederResourceRole.addToPolicy(new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        resources: ['*'],
        actions: [
            "dynamodb:GetItem",
            "dynamodb:BatchGetItem",
            "dynamodb:Query",
            "dynamodb:PutItem",
            "dynamodb:UpdateItem",
            "dynamodb:DeleteItem",
            "dynamodb:BatchWriteItem",
            "dynamodb:updateTable",
            "dynamodb:describeTable"
        ]
    }));

    const seedLayer = new lambda.LayerVersion(this, `${stackId}_SeedLayer`, {
        code: new lambda.AssetCode(`seeds`)
    })

    table.grantReadWriteData(seederResourceRole);
    const onEvent = new lambda.Function(this, `${stackId}-${tableName}-seedHandler`, {
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'seeder.handler',
        code: lambda.Code.fromAsset(path.resolve(__dirname, '../seeder')),
        environment: {
            SEED_STAGE: seedsStage,
            TABLE_NAME: tableName
        },
        layers: [seedLayer],
        timeout: cdk.Duration.minutes(15),
        role: seederResourceRole
    })
    table.grantReadWriteData(onEvent);

    const seederProvider = new cr.Provider(this, `${stackId}_seederProvider`, {
        onEventHandler: onEvent,
        logRetention: logs.RetentionDays.ONE_DAY
    })

    /**
     * NOTE: For some reason the custom resource is not triggering
     * the "Update" action for Custom Resources. You will need
     * to destroy and deploy for changes in seed data to be reflected.
     */
    new cdk.CustomResource(this, `${stackId}_seederCustomResource`, {
        serviceToken: seederProvider.serviceToken
    })

  }
}