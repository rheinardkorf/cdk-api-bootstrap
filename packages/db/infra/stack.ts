import * as cdk from "@aws-cdk/core";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import * as path from 'path'
import * as fs from 'fs'

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || "dev";
const stackId = `${wsConfig.stackPrefix}DB-${stage}`;
const tableName = `${wsConfig.defaultTable}-${stage}`;

export class ServiceStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    
    const removalPolicy = !["prod", "production"].includes(stage)
      ? cdk.RemovalPolicy.DESTROY
      : cdk.RemovalPolicy.RETAIN;

    // DynamoDB Table Definition.
    const table = new dynamodb.Table(this, `${stackId}_DynamoDB`, {
      partitionKey: { name: "PK", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK", type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      tableName,
      removalPolicy,
      stream: dynamodb.StreamViewType.NEW_AND_OLD_IMAGES,
      pointInTimeRecovery: true
    });

    // Add GSI1.
    table.addGlobalSecondaryIndex({
      indexName: "GSI1",
      partitionKey: { name: "PK1", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK1", type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.ALL,
    });

    // Add GSI2.
    table.addGlobalSecondaryIndex({
      indexName: "GSI2",
      partitionKey: { name: "PK2", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK2", type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.ALL,
    });

    // Add GSI3.
    table.addGlobalSecondaryIndex({
      indexName: "GSI3",
      partitionKey: { name: "PK3", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK3", type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.ALL,
    });
  }
}
