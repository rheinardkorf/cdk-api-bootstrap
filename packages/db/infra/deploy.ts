#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from '@aws-cdk/core';
import { ServiceStack } from './stack';
import * as path from 'path';
import * as fs from 'fs';

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || 'dev';
const stackId = `${wsConfig.stackPrefix}DB-${stage}`;
const account = process.env.CDK_DEFAULT_ACCOUNT;
const region = process.env.CDK_DEFAULT_REGION || 'ap-southeast-2';

const app = new cdk.App();
new ServiceStack(app, stackId, {
  env: {
    account,
    region,
  },
  description: `Deploying Stack (${stackId})`
});