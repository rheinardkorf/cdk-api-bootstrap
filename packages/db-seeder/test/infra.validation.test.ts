import { SynthUtils } from "@aws-cdk/assert"
import { Stack } from "@aws-cdk/core"

import { SeederStack } from "../infra/stack"

test("DB Seeder stack validation", () => {
    const stack = new Stack();
    const seeder_stack = new SeederStack(stack, 'seeder_stack');
    const received = SynthUtils.toCloudFormation(seeder_stack);
    expect(SynthUtils.toCloudFormation(seeder_stack)).toMatchSnapshot();
});