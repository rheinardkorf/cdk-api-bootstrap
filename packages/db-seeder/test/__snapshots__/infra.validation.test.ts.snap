// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DB Seeder stack validation 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48cArtifactHash0CC9DBB5": Object {
      "Description": "Artifact hash for asset \\"190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48c\\"",
      "Type": "String",
    },
    "AssetParameters190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48cS3BucketB9C082A8": Object {
      "Description": "S3 bucket for asset \\"190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48c\\"",
      "Type": "String",
    },
    "AssetParameters190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48cS3VersionKey2E07B5C3": Object {
      "Description": "S3 key for asset version \\"190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48c\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cArtifactHash627DAAA7": Object {
      "Description": "Artifact hash for asset \\"c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c\\"",
      "Type": "String",
    },
    "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3BucketEAC9DD43": Object {
      "Description": "S3 bucket for asset \\"c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c\\"",
      "Type": "String",
    },
    "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3VersionKeyDD9AE9E7": Object {
      "Description": "S3 key for asset version \\"c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c\\"",
      "Type": "String",
    },
    "AssetParametersded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eafArtifactHash20BCD136": Object {
      "Description": "Artifact hash for asset \\"ded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eaf\\"",
      "Type": "String",
    },
    "AssetParametersded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eafS3BucketC598B87E": Object {
      "Description": "S3 bucket for asset \\"ded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eaf\\"",
      "Type": "String",
    },
    "AssetParametersded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eafS3VersionKey7A068F6F": Object {
      "Description": "S3 key for asset version \\"ded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eaf\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyProjectDBSeederdevSeedLayer106D9E25": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eafS3BucketC598B87E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eafS3VersionKey7A068F6F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersded02fc38ccee7624ccf76c813fd53a5ca0ea644dc4b39b48e598ec2fb581eafS3VersionKey7A068F6F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "MyProjectDBSeederdevdatadevseedHandler01709C6C": Object {
      "DependsOn": Array [
        "MyProjectDBSeederdevseederResourceRoleDefaultPolicy7B61D9F1",
        "MyProjectDBSeederdevseederResourceRoleC51146EC",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48cS3BucketB9C082A8",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48cS3VersionKey2E07B5C3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters190e72aa1cf240afd7df504eeb8214dd94dbc2bec52d4e2cec73f367f0fbd48cS3VersionKey2E07B5C3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "SEED_STAGE": "dev",
            "TABLE_NAME": "data-dev",
          },
        },
        "Handler": "seeder.handler",
        "Layers": Array [
          Object {
            "Ref": "MyProjectDBSeederdevSeedLayer106D9E25",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyProjectDBSeederdevseederResourceRoleC51146EC",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyProjectDBSeederdevseederCustomResource": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "MyProjectDBSeederdevseederProviderframeworkonEventC1A298F9",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "MyProjectDBSeederdevseederProviderframeworkonEventC1A298F9": Object {
      "DependsOn": Array [
        "MyProjectDBSeederdevseederProviderframeworkonEventServiceRoleDefaultPolicy729EB57C",
        "MyProjectDBSeederdevseederProviderframeworkonEventServiceRole8C070508",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3BucketEAC9DD43",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3VersionKeyDD9AE9E7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49cS3VersionKeyDD9AE9E7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/seeder_stack/MyProjectDB-Seeder-dev_seederProvider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "MyProjectDBSeederdevdatadevseedHandler01709C6C",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyProjectDBSeederdevseederProviderframeworkonEventServiceRole8C070508",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyProjectDBSeederdevseederProviderframeworkonEventLogRetention4AFFF7BA": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "MyProjectDBSeederdevseederProviderframeworkonEventC1A298F9",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "MyProjectDBSeederdevseederProviderframeworkonEventServiceRole8C070508": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyProjectDBSeederdevseederProviderframeworkonEventServiceRoleDefaultPolicy729EB57C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "MyProjectDBSeederdevdatadevseedHandler01709C6C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyProjectDBSeederdevseederProviderframeworkonEventServiceRoleDefaultPolicy729EB57C",
        "Roles": Array [
          Object {
            "Ref": "MyProjectDBSeederdevseederProviderframeworkonEventServiceRole8C070508",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyProjectDBSeederdevseederResourceRoleC51146EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MyProjectDBSeederdevseederResourceRoleDefaultPolicy7B61D9F1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:GetItem",
                "dynamodb:BatchGetItem",
                "dynamodb:Query",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:updateTable",
                "dynamodb:describeTable",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/data-dev",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/data-dev/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyProjectDBSeederdevseederResourceRoleDefaultPolicy7B61D9F1",
        "Roles": Array [
          Object {
            "Ref": "MyProjectDBSeederdevseederResourceRoleC51146EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
